wire-grpc-client/src/main/java/com/squareup/wire/GrpcMethod.kt

```diff
@@ -211,6 +220,154 @@ internal sealed class GrpcMethod<S : Any, R : Any>(
+          val requestBody = newRequestBody(requestAdapter, request)
+          call = grpcClient.newCall(path, requestBody)
+
+          val messageSource = BlockingMessageSource(call!!, responseAdapter)
+          call!!.enqueue(messageSource.readFromResponseBodyCallback())
+
+          messageSource.use {
+            try {
+              val response =
+                  messageSource.read() ?: throw ProtocolException("required message not sent")
+              callback.onSuccess(this, response)
+            } catch (e: IOException) {
+              callback.onFailure(this, e)
+            }
+          }
+        }
+      }
+    }
+  }
+
+  class GrpcStreamingCall<S : Any, R : Any>(
+    path: String,
+    requestAdapter: ProtoAdapter<S>,
+    responseAdapter: ProtoAdapter<R>
+  ) : GrpcMethod<S, R>(path, requestAdapter, responseAdapter) {
+    override fun invoke(
+      grpcClient: GrpcClient,
+      args: Array<Any>
+    ): com.squareup.wire.GrpcStreamingCall<S, R> {
+      val requestBody = newDuplexRequestBody()
+      val call = grpcClient.newCall(super.path, requestBody)
+
+      return object : com.squareup.wire.GrpcStreamingCall<S, R> {
+        override val timeout: Timeout
+          get() = call.timeout()
+
+        override fun cancel() {
+          call.cancel()
+        }
+
+        override fun execute(): Pair<SendChannel<S>, ReceiveChannel<R>> {
+          // TODO(benoit) This is similar to what `callWithChannels(grpcClient)` does so we can try
+          //  to share; but I didn't find a way to call `super@com.squareup.wire.GrpcStreamingCall`.
+          val requestChannel = Channel<S>(1)
+          val responseChannel = Channel<R>(1)
+          requestChannel.writeToRequestBody(requestBody, requestAdapter, call)
+          call.enqueue(responseChannel.readFromResponseBodyCallback(responseAdapter))
+
+          responseChannel.invokeOnClose { cause ->
+            call.cancel()
+            requestChannel.cancel()
+            responseChannel.cancel()
+          }
+
+          return requestChannel to responseChannel
+        }
+
+        override fun executeBlocking(): Pair<MessageSink<S>, MessageSource<R>> {
+          val messageSource = BlockingMessageSource(call, responseAdapter)
+          val messageSink = requestBody.messageSink(requestAdapter, call)
+          call.enqueue(messageSource.readFromResponseBodyCallback())
+
+          return messageSink to messageSource
+        }
+
+        override fun enqueue(callback: com.squareup.wire.GrpcStreamingCall.Callback<S, R>) {
+          val messageSource = BlockingMessageSource(call, responseAdapter)
+          val messageSink = requestBody.messageSink(requestAdapter, call)
+          try {
+            call.enqueue(messageSource.readFromResponseBodyCallback())
+            callback.onSuccess(this, messageSink, messageSource)
+          } catch (e: IOException) {
+            // TODO(benoit) :no-idea-what-im-doing:
```

Comment:
_almost_ right. I’d prefer to guard this a bit:

```
          var calledCallback = false
          try {
            call.enqueue(messageSource.readFromResponseBodyCallback())
            calledCallback = true
            callback.onSuccess(this, messageSink, messageSource)
          } catch (e: IOException) {
            if (!calledCallback) {
             callback.onFailure(this, e) 
            }
          }
```

That way if you throw in onSuccess() you don’t get called on onFailure().